chroma_api/stratagem.py

```diff
@@ -0,0 +1,145 @@
+from django.db import models
+from django.db.models import Q
+from django.core.exceptions import ObjectDoesNotExist
+from toolz.functoolz import pipe
+
+import tastypie.http as http
+
+from tastypie import fields
+from tastypie.resources import Resource
+from chroma_api.authentication import AnonymousAuthentication
+from chroma_core.services.job_scheduler.job_scheduler_client import JobSchedulerClient
+from tastypie.authorization import DjangoAuthorization
+from tastypie.validation import Validation
+from chroma_api.validation_utils import validate
+from chroma_api.utils import custom_response, dehydrate_command
+from chroma_core.models import (
+    StratagemConfiguration,
+    ManagedHost,
+    ManagedMdt,
+    ManagedTargetMount,
+    ManagedFilesystem,
+    Command,
+)
+
+from chroma_api.chroma_model_resource import ChromaModelResource
+
+
+class RunStratagemValidation(Validation):
+    def is_valid(self, bundle, request=None):
+        purge_duration = bundle.data.get("purge_duration")
+        report_duration = bundle.data.get("report_duration")
+
+        if "filesystem" not in bundle.data:
+            return {"code": "filesystem_required", "message": "filesystem required when running stratagem."}
+        elif isinstance(purge_duration, int) and isinstance(report_duration, int) and report_duration >= purge_duration:
+            return {"code": "duration_order_error", "message": "Report duration must be less than purge duration."}
+
+        fs_identifier = str(bundle.data.get("filesystem"))
+
+        if not any(
+            map(
+                lambda x: str(x.get("id")) == fs_identifier or str(x.get("name")) == fs_identifier,
+                ManagedFilesystem.objects.values("id", "name"),
+            )
+        ):
+            return {
+                "code": "filesystem_does_not_exist",
+                "message": "Filesystem {} does not exist.".format(fs_identifier),
+            }
+
+        fs_id = filter(
+            lambda x: str(x.get("id")) == fs_identifier or str(x.get("name")) == fs_identifier,
+            ManagedFilesystem.objects.values("id", "name"),
+        )[0].get("id")
+
+        # Each MDT associated with the fielsystem must be installed on a server that has the stratagem profile installed
+        target_mount_ids = map(lambda mdt: mdt.active_mount_id, ManagedMdt.objects.filter(filesystem_id=fs_id))
```

Comment:
```suggestion
        target_mount_ids = ManagedMdt.objects.filter(filesystem_id=1).values('active_mount_id')
```

Reply:
Note: the `active_mount_id`s may be `None`

Reply:
I added:
```
[x for x in target_mount_ids if x is not None]
```
To filter out any `None` values.

Reply:
What happens if all targets are unmounted?

Reply:
I'm just think about this a bit more... If the target is unmounted then it's basically not going to be able to run stratagem. Would it be acceptable to say that if "None" is returned for any of these then return an error indicating that all mdt targets must be mounted?

Reply:
Based on our discussion i've updated this to the following:
```
        target_mount_ids = (
            ManagedMdt.objects.filter(filesystem_id=fs_id, active_mount_id__isnull=False)
            .values_list("active_mount_id", flat=True)
            .distinct()
        )
        mdt0 = ManagedMdt.objects.filter(filesystem_id=fs_id, name__contains="MDT0000").first()

        if mdt0 is None:
            return {"code": "mdt0_not_found", "message": "MDT0 could not be found."}

        if mdt0.active_mount_id not in target_mount_ids:
            return {"code": "mdt0_not_mounted", "message": "MDT0 must be mounted in order to run stratagem."}
```